### PR TITLE
Rewrite the validators update part of global-state-update-gen

### LINKED ISSUE #2904: Fix the total supply when updating validators in global-state-update-gen
The global-state-update gen tool currently only alters bids when modifying the validator set, without taking care of the bid purse or the total token supply. It should be fixed so that it keeps all this state consistent.
We should also add tests for this

Reported by @fizyk20

### PR DESCRIPTION
**Pull request by fizyk20**
These changes make sure that:
- The bonding purses for bids aren't the same as the validators' main purses, and that they contain amounts of tokens consistent with the bids.
- The tokens are transferred instead of created/burned where applicable.
- The balances of bonding purses of the validators being removed are set to 0.
- The total supply is being correctly tracked.

Closes #2904

### CODE REVIEW

**Review by mpapierski — APPROVED** [review]
Haven't tested the tool, but the logic looks solid. Each time a bid is modified, a bonding purse, and optionally validator's main purse is also adjusted and amounts minted/burned are reflected in total supply as well

File: utils/global-state-update-gen/src/validators/validators_manager.rs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        &mut self,
+        validators_diff: &ValidatorsDiff,
+        new_snapshot: &SeigniorageRecipientsSnapshot,
+    ) {
+        let large_bids = self.find_large_bids(new_snapshot);
+        let to_unbid = validators_diff.removed.union(&large_bids);
+
+        let bids = self.builder.get_bids();
+
+        for (pkey, seigniorage_recipient) in new_snapshot.values().next().unwrap() {
```
Comment (Fraser999) [inline utils/global-state-update-gen/src/validators/validators_manager.rs]: `pkey` to `pub_key` or `public_key` pls throughout this PR.

**Review by Fraser999 — APPROVED** [review]
One nitpick, but otherwise LGTM.

**Comment by fizyk20** [comment]
bors merge

**Pull request merged**